datafusion/physical-expr/src/datetime_expressions.rs

```diff
@@ -662,8 +662,34 @@ fn _date_trunc_coarse_with_tz(
     granularity: &str,
     value: Option<DateTime<Tz>>,
 ) -> Result<Option<i64>> {
-    let value = _date_trunc_coarse::<DateTime<Tz>>(granularity, value)?;
-    Ok(value.and_then(|value| value.timestamp_nanos_opt()))
+    if let Some(value) = value {
+        let local = value.naive_local();
+        let truncated = _date_trunc_coarse::<NaiveDateTime>(granularity, Some(local))?;
+        let truncated = truncated.map(|truncated| {
+            match truncated.and_local_timezone(value.timezone()) {
+                LocalResult::None => {
+                    // It is impossible to truncate from a time that does exist into one that doesn't.
+                    panic!("date_trunc produced impossible time")
```

Comment:
Panic! seems rather severe for this - exec_err! I think would be a better option as it will bubble up the error back to the caller.